### ISSUE TITLE
[ML/Dynamics] Add bounded Training/Evaluation session semantics with Host-owned autodiff and optimizer internals

### ISSUE THREAD

**Issue by dancxjo** [dancxjo/conduit#2140]
Parent: #2132
Depends on: #2133 #2137 #2138 #2139

## Outcome

Add provider-neutral training/evaluation semantics that let a Conduit Form run bounded model-learning work, update explicit model/training state through #2133 state boundaries, record objectives/metrics/provenance, and emit durable checkpoints, while leaving autograd graphs, tensor allocators, optimizer kernels, mixed precision, and accelerator-specific machinery inside the selected Host realization.

## Background

The Tongues proof needs more than inference. It needs to train a shared latent model from paired and partially paired audio/articulatory observations using several simultaneous objectives.

Conceptually:

```text
checkpoint/model-state[n]
        |
        v
      batch ---> model ---> predictions
        |                    |
        +------ targets -----+
                             v
                           losses
                             |
                             v
                        train step
                             |
                             v
                   model-state[n+1]
                             |
                    explicit State boundary
```

The wrong implementation would expose framework-internal operations as thousands of Conduit Gears (`matmul`, `backward`, `AdamW`, etc.) before there is a demonstrated need.

The right first implementation treats one admitted training/evaluation step as a semantic Host capability with exact model/data/objective/work identities and bounded results.

## Semantic concepts

Define finite, provider-neutral concepts equivalent to:

### TrainingRun / TrainingSession descriptor

Identifies:

- base model artifact/checkpoint;
- dataset/split identity;
- training profile/objective configuration;
- stochastic seed/profile;
- requested work bounds;
- selected precision/profile where semantically material;
- checkpoint cadence/policy if any;
- evaluation policy if any.

### TrainingState

Explicit mutable state needed between steps, distinguishable from immutable model artifacts/checkpoints. It may include model parameters plus provider-defined optimizer/scaler state behind a compatible semantic state schema.

### TrainStep

Consumes admitted model/training state + one bounded batch/example set and produces:

- next training state;
- bounded loss/metric report;
- work/accounting report;
- terminal/refusal state.

### Evaluate

Runs the same/related model without permitted parameter update and returns metrics/predictions/provenance.

### Checkpoint

Materializes a durable immutable checkpoint identity from current admitted state; compatible with #2137.

## Framework boundary

Host adapters may use:

- PyTorch;
- JAX;
- Burn;
- Candle;
- another runtime.

Inside that Host operation they may own:

- automatic differentiation tape;
- framework tensors;
- graph compilation;
- optimizer implementation;
- gradient scaling;
- mixed precision;
- CUDA/ROCm/Metal/etc. kernels;
- gradient checkpointing;
- temporary device allocations.

Those implementation details must not become mandatory semantic objects merely because one adapter uses them.

Conduit owns:

- exact model/checkpoint identity;
- exact dataset/batch identity;
- objective/profile identity;
- state generation;
- work/resource admission;
- seed/stochastic provenance;
- result/checkpoint identity;
- observed metrics and terminal disposition.

## Objective/loss semantics

Support a bounded objective descriptor capable of representing the Tongues experiment without requiring Conduit to implement the arithmetic itself.

Needed objective roles include at least:

- acoustic reconstruction;
- articulatory reconstruction;
- acoustic -> articulatory cross-modal reconstruction;
- articulatory -> acoustic cross-modal reconstruction;
- paired latent agreement/alignment;
- probabilistic/KL or comparable regularization where selected;
- optional cycle/reconstruction consistency;
- regularization terms;
- weighted total objective.

A generic objective component should identify:

- semantic role/name;
- finite weight/configuration;
- metric/loss output identity;
- optimization participation vs observation-only metric.

Do not canonize one specific VAE loss formulation as architecture truth.

## Important inverse-problem constraint

Do **not** blindly require articulatory cycle output to equal the exact original articulation when acoustics underdetermine articulation.

For a path like:

```text
A -> audio -> inferred A'
```

exact `A' == A` may be scientifically wrong if multiple articulations are acoustically equivalent.

Training/evaluation profiles need to permit probabilistic/plausibility or acoustically equivalent consistency objectives rather than forcing false uniqueness.

## Batching/data access

Training batches should consume exact example/resource identities from #2139.

Requirements:

- bounded batch size/items/bytes;
- split membership validation hooks;
- deterministic/shuffled sampling profile with explicit seed where applicable;
- missing-modality handling according to declared model/objective support;
- no accidental ambient filesystem sweep for “whatever training data exists.”

## State and recurrence

Use #2133 rather than arbitrary cyclic Forms:

```text
State<TrainingState>[n]
        -> TrainStep
        -> next TrainingState
        -> commit
        -> State<TrainingState>[n+1]
```

The planner must still see an acyclic graph per training step.

## Work/resource accounting

Training is substantially heavier than existing local inference. Admission should cover at least:

- model/checkpoint bytes;
- working memory/device memory ceiling;
- accelerator/compute lanes or equivalent resource class;
- maximum batch bytes/items;
- maximum step count/work units;
- checkpoint output bytes;
- in-flight/queue limits;
- cancellation support.

Training adapters must fail/refuse cleanly on OOM/resource exhaustion rather than silently exceeding Plan bounds.

## Lifecycle/cancellation

Define behavior for:

- load/initialize;
- ready;
- active step;
- evaluation;
- checkpointing;
- cancellation;
- provider loss;
- terminal failure;
- clean unload.

Cancellation must not ambiguously commit partially updated model state. Either the step commits atomically at the semantic boundary or returns an explicit partial/failed state profile if such a mode is deliberately admitted.

## Reproducibility/provenance

A training/checkpoint receipt should be able to identify:

- base model/checkpoint;
- dataset manifest and split;
- exact preprocessing/derived observation identities;
- objective profile;
- seed/randomness profile;
- runtime/adapter/version;
- hardware/precision profile where material;
- admitted resource envelope;
- completed step range/count;
- loss/metric summaries;
- output checkpoint identity.

Do not promise bitwise reproducibility where the selected realization cannot provide it.

## Non-goals

- no Conduit-native autograd engine;
- no primitive semantic Gear for every optimizer/tensor operation;
- no global mutable model state outside #2133-style explicit state;
- no unbounded epochs/queues/checkpoints;
- no framework-specific checkpoint path as portable identity;
- no false claim that deterministic seed implies bitwise equality on nondeterministic accelerator kernels.

## Acceptance criteria

- [ ] Provider-neutral TrainingSession/TrainStep/Evaluate/Checkpoint contracts exist.
- [ ] TrainStep consumes explicit state and produces next state; recurrence is expressed through #2133.
- [ ] Host adapter owns autodiff/optimizer/tensor internals.
- [ ] Objective configuration supports multiple weighted loss/metric components.
- [ ] Training can use paired and missing-modality examples from #2139 under explicit profile rules.
- [ ] Resource admission includes model memory, working memory/compute, batch bounds, step/work bounds, and checkpoint output.
- [ ] Cancellation/failure cannot silently commit an indeterminate half-step.
- [ ] Training provenance ties output checkpoint to base model, data split, objectives, seed, realization, and work.
- [ ] Evaluation cannot mutate model state.
- [ ] A proof runs several bounded update steps and emits a new immutable checkpoint identity.

## Architectural test

> **Conduit can say exactly what was trained, on what evidence, under what bounds, and what checkpoint resulted, without needing to know how PyTorch represented the backward graph.**

**Comment by dancxjo** [comment]
Starting from exact main `dfafa7a8aa94e82109d760da785e421db114ec92`. Outcome: provider-neutral bounded training/evaluation/checkpoint semantics and multi-step proof. Non-goals: framework internals, adapter implementation, cyclic Forms, product UI. Initial ownership is `semantics/ai` modules/tests; exact file list will stay within that crate unless a dependency gap is reported.

**Comment by dancxjo** [comment]
Accepted on exact `main` commit `ef93fb6f47dd53f5bafff1c4720b6b21a3263dba` via #2191.

Exact PR-head evidence:

- head `fa59966565677a095c24cae25d7e4a729bb7b9cc`
- required `check` passed: https://github.com/dancxjo/conduit/actions/runs/33703310680
- all six selected workspace shards passed
- the real classifier demonstration skipped unrelated ESP32 firmware and heavyweight ConduitOS/browser product jobs while preserving the stable required gates

Exact-main focused acceptance:

`cargo +stable test -p conduit-ai --locked --test training --test training_lifecycle`

Result: 4 passed, 0 failed at `ef93fb6f47dd53f5bafff1c4720b6b21a3263dba`.

This establishes the finite provider-neutral session/step/evaluation/checkpoint contracts, explicit #2133 recurrence, Host-owned computational internals, bounded paired/missing-modality data and resources, atomic failure/cancellation behavior, immutable checkpoint provenance, non-mutating evaluation, lifecycle terminals, and the several-step proof. #2132 remains open; #2141 onward are not claimed here.
